scipy/sparse/csgraph/_flow.pyx

```diff
@@ -668,3 +682,753 @@ cdef ITYPE_t[:] _dinic(
+                             shape=(n_verts, n_verts))
+    return MinCostFlowResult(flow_value, flow_matrix,
+                             flow_cost)
+
+
+def _network_simplex_checks(
+        capacities,  # IN
+        demand,  # IN
+        cost,  # IN
+        ):
+    if np.any(demand == INT_MAX) or np.any(demand == INT_MIN):
+        raise ValueError("one of the vertices has infinite demand")
+
+    if np.any(cost == INT_MAX) or np.any(cost == INT_MIN):
+        raise ValueError("one of the edges has infinite cost")
+
+    if np.any(capacities < 0):
+        raise ValueError("one of the edges has negative capacity")
+
+    demand_sum = demand.sum()
+
+    if demand_sum != 0:
+        raise ValueError("sum of demands is not zero")
+
+
+cdef void _initialize_spanning_tree(
+        ITYPE_t n_verts,  # IN
+        ITYPE_t n_non_zero_edges,  # IN
+        ITYPE_t faux_inf,  # IN
+        ITYPE_t[:] demand,  # IN
+        ITYPE_t[:] edge_flow,  # IN/OUT
+        ITYPE_t[:] vertex_potentials,  # IN/OUT
+        ITYPE_t[:] parent,  # IN/OUT
+        ITYPE_t[:] parent_edge,  # IN/OUT
+        ITYPE_t[:] subtree_size,  # IN/OUT
+        ITYPE_t[:] next_vertex_dft,  # IN/OUT
+        ITYPE_t[:] prev_vertex_dft,  # IN/OUT
+        ITYPE_t[:] last_descendent_dft  # IN/OUT
+        ) nogil:
+    cdef ITYPE_t e, v
+
+    # edge_flow initialization
+    for e in range(n_non_zero_edges):
+        edge_flow[e] = 0
+    for v in range(n_verts):
+        edge_flow[v + n_non_zero_edges] = abs(demand[v])
+
+    # vertex_potentials initialization
+    for v in range(1, n_verts + 1):
+        if demand[v - 1] <= 0:
+            vertex_potentials[v] = faux_inf
+        else:
+            vertex_potentials[v] = -faux_inf
+
+    # parent initialization
+    parent[0] = -2
+    subtree_size[0] = n_verts + 1
+    prev_vertex_dft[0] = n_verts
+    last_descendent_dft[0] = n_verts
+    for v in range(1, n_verts + 1):
+        parent[v] = 0
+        parent_edge[v] = n_non_zero_edges + v - 1
+        subtree_size[v] = 1
+        prev_vertex_dft[v] = v - 1
+        last_descendent_dft[v] = v
+
+    # next_vertex_dft initialization
+    next_vertex_dft[0] = 1
+    for v in range(1, n_verts):
+        next_vertex_dft[v] = v + 1
+    next_vertex_dft[n_verts] = 0
+
+
+cdef inline ITYPE_t _reduced_cost(
+        ITYPE_t e,  # IN
+        ITYPE_t[:] edge_weights,  # IN
+        ITYPE_t[:] vertex_potentials,  # IN
+        ITYPE_t[:] edge_sources,  # IN
+        ITYPE_t[:] edge_targets,  # IN
+        ITYPE_t[:] edge_flow  # IN
+        ) nogil:
+    cdef ITYPE_t c
+    c = (edge_weights[e]
+         - vertex_potentials[edge_sources[e]]
+         + vertex_potentials[edge_targets[e]])
+    return c if edge_flow[e] == 0 else -c
+
+ctypedef struct local_vars:
+    ITYPE_t B
+    ITYPE_t M
+    ITYPE_t m
+    ITYPE_t f
+
+ctypedef struct network_simplex_result:
+    ITYPE_t flow_cost
+    ITYPE_t size
+    ITYPE_t flow_value
+    bint is_correct
```

Comment:
If you need to return such struct to Python, the best is to use [Cython extension types](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#extension-types) with [`readonly`](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#static-attributes)-qualified fields.

Simple definition of extension types simply adds the [`PyObject_HEAD`](https://docs.python.org/3/c-api/structures.html#c.PyObject_HEAD) macro at the start of the struct making this objects returnable in Python. `readonly` allows accessing fields in Python which aren't by default.

Reply:
I had to change `ctypedef struct` to `cdef class` and added `readonly` attribute to every member of `network_simplex_result`. `return ns_result` should be done with `gil`. Though overall, there is no decrease in performance. In fact, there is a slight improvement observed.

Reply:
Thanks for reporting this.